query_optimizer/ExecutionGenerator.cpp

```diff
@@ -740,6 +805,202 @@ void ExecutionGenerator::convertHashJoin(const P::HashJoinPtr &physical_plan) {
+
+    switch (build_partition_scheme_header.getPartitionType()) {
+      case PartitionSchemeHeader::PartitionType::kRange:
+        build_needs_repartition = true;
+
+        switch (probe_partition_scheme_header.getPartitionType()) {
+          case PartitionSchemeHeader::PartitionType::kRange:
+            probe_needs_repartition = true;
+            break;
+          case PartitionSchemeHeader::PartitionType::kHash: {
+            const attribute_id probe_partition_attr = probe_partition_scheme_header.getPartitionAttributeId();
+            if (find(probe_attribute_ids.begin(), probe_attribute_ids.end(), probe_partition_attr) !=
+                    probe_attribute_ids.end()) {
+              needs_swap = true;
+            } else {
+              probe_needs_repartition = true;
+            }
+            break;
+          }
+        }
+        break;
+      case PartitionSchemeHeader::PartitionType::kHash: {
+        const attribute_id build_partition_attr = build_partition_scheme_header.getPartitionAttributeId();
+        if (find(build_attribute_ids.begin(), build_attribute_ids.end(), build_partition_attr) !=
+                build_attribute_ids.end()) {
+          // BuildRelation has a useful partition.
+          switch (probe_partition_scheme_header.getPartitionType()) {
+            case PartitionSchemeHeader::PartitionType::kRange:
+              probe_needs_repartition = true;
+              break;
+            case PartitionSchemeHeader::PartitionType::kHash: {
+              if (areSamePartitionSchemeHeaders(build_partition_scheme_header, *build_relation,
+                                                probe_partition_scheme_header, *probe_relation)) {
+                if (cost_model_for_hash_join_->estimateCardinality(build_physical) >
+                        cost_model_for_hash_join_->estimateCardinality(probe_physical)) {
+                  needs_swap = true;
+                }
+              } else {
+                probe_needs_repartition = true;
+              }
+              break;
+            }
+          }
+        } else {
+          build_needs_repartition = true;
+
+          switch (probe_partition_scheme_header.getPartitionType()) {
+            case PartitionSchemeHeader::PartitionType::kRange:
+              probe_needs_repartition = true;
+              break;
+            case PartitionSchemeHeader::PartitionType::kHash: {
+              const attribute_id probe_partition_attr = probe_partition_scheme_header.getPartitionAttributeId();
+              if (find(probe_attribute_ids.begin(), probe_attribute_ids.end(), probe_partition_attr) !=
+                      probe_attribute_ids.end()) {
+                needs_swap = true;
+              } else {
+                probe_needs_repartition = true;
+              }
+              break;
+            }
+          }
+        }
+        break;
+      }
+    }
```

Comment:
I would highly recommend two things for the switch statement:
1. Add several comments in between to describe what's happening. I felt lost while reading it.
2. Moving the switch statement to a separate method.